File: electron/resultados.js
'use strict';

/**
 * Onde os resultados são gravados, e o que se apaga sozinho.
 *
 * Tudo vai para `Downloads/PDF.GreenCodes`. Downloads porque é onde a pessoa
 * já procura arquivo baixado; a subpasta porque a auto-exclusão varre por
 * tempo, e varrer a raiz de Downloads seria apagar coisa que não é nossa.
 *
 * A auto-exclusão é opcional e por arquivo: quem marcou entra num registro
 * com a hora de morrer, e uma varredura periódica apaga o que passou do
 * prazo. O que não foi marcado fica para sempre — é arquivo da pessoa, no
 * computador dela.
 */

const fsp = require('node:fs/promises');
const path = require('node:path');

const UM_DIA = 24 * 60 * 60 * 1000;
const INTERVALO_DA_VARREDURA = 60 * 60 * 1000;

class Resultados {
  /**
   * @param {string} pastaDeDownloads onde gravar
   * @param {string} pastaDeDados onde guardar o registro da auto-exclusão
   */
  constructor(pastaDeDownloads, pastaDeDados) {
    this.pasta = path.join(pastaDeDownloads, 'PDF.GreenCodes');
    this.registro = path.join(pastaDeDados, 'auto-exclusao.json');
    this.timer = null;
  }

  async garantirPasta() {
    await fsp.mkdir(this.pasta, { recursive: true });
    return this.pasta;
  }

  /**
   * Grava com o primeiro número livre: 1.pdf, 2.pdf, 3.pdf.
   *
   * Sem diálogo e sem sobrescrever nada. Quem processa vários documentos
   * seguidos não quer decidir nome e pasta a cada um.
   */
  async salvar(nome, bytes, { apagarEm1Dia = false } = {}) {
    await this.garantirPasta();

    const extensao = path.extname(nome) || '.pdf';
    const existentes = new Set(await fsp.readdir(this.pasta));
    let numero = 1;
    while (existentes.has(numero + extensao)) numero += 1;

    const destino = path.join(this.pasta, numero + extensao);
    await fsp.writeFile(destino, Buffer.from(bytes));

    if (apagarEm1Dia) await this.marcarParaApagar(destino);
    return destino;
  }

  async lerRegistro() {
    try {
      const conteudo = JSON.parse(await fsp.readFile(this.registro, 'utf8'));
      return Array.isArray(conteudo) ? conteudo : [];
    } catch {
      // Registro ausente ou corrompido é o mesmo que registro vazio: nada
      // para apagar é sempre a resposta segura.
      return [];
    }
  }

  async gravarRegistro(itens) {
    await fsp.mkdir(path.dirname(this.registro), { recursive: true });
    await fsp.writeFile(this.registro, JSON.stringify(itens, null, 2), 'utf8');
  }

  async marcarParaApagar(caminho, prazoMs = UM_DIA) {
    const itens = (await this.lerRegistro()).filter((item) => item.caminho !== caminho);
    itens.push({ caminho, apagarEm: Date.now() + prazoMs });
    await this.gravarRegistro(itens);
  }

  /** Tira a marca sem apagar o arquivo. */
  async manterParaSempre(caminho) {
    const itens = await this.lerRegistro();
    await this.gravarRegistro(itens.filter((item) => item.caminho !== caminho));
  }

  /**
   * Apaga o que passou do prazo e limpa o registro.
   *
   * Some do registro também o que já não existe — arquivo que a pessoa moveu
   * ou apagou à mão não deveria ficar sendo procurado para sempre.
   */
  async varrer(agora = Date.now()) {
    const itens = await this.lerRegistro();
    if (itens.length === 0) return { apagados: 0, restantes: 0 };

    const ficam = [];
    let apagados = 0;

    for (const item of itens) {
      // Só mexe no que está dentro da nossa pasta: uma entrada estragada não
      // pode virar permissão para apagar arquivo em outro lugar.
      const dentroDaPasta = path.resolve(item.caminho).startsWith(path.resolve(this.pasta) + path.sep);
      if (!dentroDaPasta) continue;

      if (agora < item.apagarEm) {
        // Arquivo que sumiu não precisa continuar no registro.
        try {
          await fsp.access(item.caminho);
          ficam.push(item);
        } catch {
          /* já não existe */
        }
        continue;
      }

      try {
        await fsp.rm(item.caminho, { force: true });
        apagados += 1;
      } catch {
        // Arquivo aberto em outro programa não apaga agora; fica para a
        // próxima varredura em vez de sumir do registro sem ter sido apagado.
        ficam.push(item);
      }
    }

    await this.gravarRegistro(ficam);
    return { apagados, restantes: ficam.length };
  }

  /** Varre agora e de hora em hora enquanto o aplicativo estiver aberto. */
  iniciarVarredura() {
    void this.varrer();
    this.timer = setInterval(() => void this.varrer(), INTERVALO_DA_VARREDURA);
    // Sem isto o intervalo seguraria o processo vivo ao fechar a janela.
    if (this.timer.unref) this.timer.unref();
  }

  pararVarredura() {
    if (this.timer) clearInterval(this.timer);
    this.timer = null;
  }
}

module.exports = { Resultados, UM_DIA };

File: src-tauri/src/resultados.rs
//! Onde os resultados sao gravados, e o que se apaga sozinho.
//!
//! Tudo vai para `Downloads/PDF.GreenCodes`. Downloads porque e onde a pessoa
//! ja procura arquivo baixado; a subpasta porque a auto-exclusao varre por
//! tempo, e varrer a raiz de Downloads seria apagar coisa que nao e nossa.
//!
//! A auto-exclusao e opcional e por arquivo: quem marcou entra num registro
//! com a hora de morrer, e uma varredura periodica apaga o que passou do
//! prazo. O que nao foi marcado fica para sempre - e arquivo da pessoa, no
//! computador dela.

use std::fs;
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};

const UM_DIA: u64 = 24 * 60 * 60 * 1000;
const INTERVALO_DA_VARREDURA: Duration = Duration::from_secs(60 * 60);

#[derive(Serialize, Deserialize, Clone)]
struct Marcado {
    caminho: String,
    #[serde(rename = "apagarEm")]
    apagar_em: u64,
}

fn agora() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

/// A pasta dos resultados, criada se ainda nao existir.
pub fn pasta(app: &AppHandle) -> Result<PathBuf, String> {
    let base = app
        .path()
        .download_dir()
        .or_else(|_| app.path().document_dir())
        .map_err(|e| format!("nao achei a pasta de downloads: {e}"))?;
    let pasta = base.join("PDF.GreenCodes");
    fs::create_dir_all(&pasta).map_err(|e| e.to_string())?;
    Ok(pasta)
}

fn registro(app: &AppHandle) -> Result<PathBuf, String> {
    let base = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("nao achei a pasta de dados: {e}"))?;
    fs::create_dir_all(&base).map_err(|e| e.to_string())?;
    Ok(base.join("auto-exclusao.json"))
}

/// Registro ausente ou estragado e o mesmo que registro vazio: nada para
/// apagar e sempre a resposta segura.
fn ler(app: &AppHandle) -> Vec<Marcado> {
    let Ok(arquivo) = registro(app) else {
        return Vec::new();
    };
    let Ok(texto) = fs::read_to_string(arquivo) else {
        return Vec::new();
    };
    serde_json::from_str(&texto).unwrap_or_default()
}

fn gravar(app: &AppHandle, itens: &[Marcado]) -> Result<(), String> {
    let arquivo = registro(app)?;
    let texto = serde_json::to_string_pretty(itens).map_err(|e| e.to_string())?;
    fs::write(arquivo, texto).map_err(|e| e.to_string())
}

/// Grava com o primeiro numero livre: 1.pdf, 2.pdf, 3.pdf.
///
/// Sem dialogo e sem sobrescrever nada. Quem processa varios documentos
/// seguidos nao quer decidir nome e pasta a cada um.
pub fn salvar(
    app: &AppHandle,
    nome: &str,
    bytes: &[u8],
    apagar_em_1_dia: bool,
) -> Result<PathBuf, String> {
    let pasta = pasta(app)?;
    let extensao = Path::new(nome)
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| format!(".{e}"))
        .unwrap_or_else(|| ".pdf".into());

    let mut numero = 1u32;
    let destino = loop {
        let tentativa = pasta.join(format!("{numero}{extensao}"));
        if !tentativa.exists() {
            break tentativa;
        }
        numero += 1;
        if numero > 1_000_000 {
            return Err("a pasta de resultados esta cheia demais".into());
        }
    };

    crate::arquivos::gravar(&destino, bytes)?;
    if apagar_em_1_dia {
        marcar(app, &destino.to_string_lossy())?;
    }
    Ok(destino)
}

pub fn marcar(app: &AppHandle, caminho: &str) -> Result<(), String> {
    let mut itens: Vec<Marcado> = ler(app)
        .into_iter()
        .filter(|item| item.caminho != caminho)
        .collect();
    itens.push(Marcado {
        caminho: caminho.to_string(),
        apagar_em: agora() + UM_DIA,
    });
    gravar(app, &itens)
}

/// Tira a marca sem apagar o arquivo.
pub fn manter(app: &AppHandle, caminho: &str) -> Result<(), String> {
    let itens: Vec<Marcado> = ler(app)
        .into_iter()
        .filter(|item| item.caminho != caminho)
        .collect();
    gravar(app, &itens)
}

/// Apaga o que passou do prazo e limpa o registro.
///
/// Some do registro tambem o que ja nao existe - arquivo que a pessoa moveu
/// ou apagou a mao nao deveria ficar sendo procurado para sempre.
pub fn varrer(app: &AppHandle) -> (usize, usize) {
    let itens = ler(app);
    if itens.is_empty() {
        return (0, 0);
    }
    let Ok(nossa) = pasta(app) else {
        return (0, itens.len());
    };

    let mut ficam = Vec::new();
    let mut apagados = 0usize;
    let agora = agora();

    for item in itens {
        // So mexe no que esta dentro da nossa pasta: uma entrada estragada
        // nao pode virar permissao para apagar arquivo em outro lugar.
        let alvo = PathBuf::from(&item.caminho);
        if !alvo.starts_with(&nossa) {
            continue;
        }

        if agora < item.apagar_em {
            if alvo.exists() {
                ficam.push(item);
            }
            continue;
        }

        // Arquivo aberto em outro programa nao apaga agora; fica para a
        // proxima varredura em vez de sumir do registro sem ter sido apagado.
        match fs::remove_file(&alvo) {
            Ok(()) => apagados += 1,
            Err(_) if !alvo.exists() => {}
            Err(_) => ficam.push(item),
        }
    }

    let restantes = ficam.len();
    let _ = gravar(app, &ficam);
    (apagados, restantes)
}

/// Varre agora e de hora em hora enquanto o aplicativo estiver aberto.
pub fn iniciar_varredura(app: &AppHandle) {
    let app = app.clone();
    std::thread::spawn(move || loop {
        varrer(&app);
        std::thread::sleep(INTERVALO_DA_VARREDURA);
    });
}
